### PR TITLE
Add option to customize project access roles

### PR DESCRIPTION
**Pull request by rottencandy**
Issue: https://issues.redhat.com/browse/ODC-5447

Extends the `Console` CRD to add configuration option to allow cluster admins to specify and customize project access roles.

According to enhancement proposal: https://github.com/openshift/enhancements/pull/668

### CODE REVIEW

**Comment by jerolimov** [comment]
Lgtm! Exactly how we have defined such fields the last time.

Question to go expert @jhadvig, similar to the last time we don't use a pointer ref here, right? A pointer makes it in my understanding easier to differs later between "not configured" (key is missing) and an empty list.

Or do we go the go standard way here where don't use a pointer and we don't need to check for nil later?

/cc
/uncc abhinavdahiya derekwaynecarr
/assign spadgett jhadvig

**Review by jerolimov — APPROVED** [review]

File: operator/v1/types_console.go [hunk 1]
```diff
@@ -113,6 +113,19 @@ type ConsoleCustomization struct {
 	// +kubebuilder:validation:Optional
 	// +optional
 	DeveloperCatalog DeveloperConsoleCatalogCustomization `json:"developerCatalog,omitempty"`
+	// projectAccess defines configuration options for defining project access roles.
```
Comment (jhadvig) [inline operator/v1/types_console.go:116]: This description is a bit vague, I would rather make it clear what options are those

File: operator/v1/types_console.go [hunk 2]
```diff
@@ -113,6 +113,19 @@ type ConsoleCustomization struct {
...
+	ProjectAccess ProjectAccess `json:"projectAccess,omitempty"`
+}
+
+// ProjectAccess contains options for project access roles
+type ProjectAccess struct {
+	// availableClusterRoles specifies the list of ClusterRoles that are assignable to users
+	// through the project access tab.
+	// +kubebuilder:validation:Optional
+	// +optional
+	AvailableClusterRoles []string `json:"availableClusterRoles"`
```
Comment (jhadvig) [inline operator/v1/types_console.go:128]: do we need empty list here ? if not then lets add 
```suggestion
	AvailableClusterRoles []string `json:"availableClusterRoles, omitempty"`
```

**Review by jhadvig — COMMENTED** [review]
>  A pointer makes it in my understanding easier to differs later between "not configured" (key is missing) and an empty list. 

@jerolimov yes

> Or do we go the go standard way here where don't use a pointer and we don't need to check for nil later?

so it depends on the fields semantics. Do we wanna distinguish between empty `projectAccess` field and not set at all? Not sure if that makes sense in this case. Would rather do the same as we did with `DeveloperCatalog`. In that case if the `projectAccess` field is empty it will just be omitted.

File: operator/v1/types_console.go (hunk 2 above)
Comment (rottencandy) [reply operator/v1/types_console.go]: Updated.

File: operator/v1/types_console.go (hunk 1 above)
Comment (rottencandy) [reply operator/v1/types_console.go]: Updated to describe the option.

**Review by jhadvig — APPROVED** [review]
Thanks 👍 we will need a final review from @spadgett here.

File: operator/v1/types_console.go [hunk 3]
```diff
@@ -113,6 +113,21 @@ type ConsoleCustomization struct {
...
+	// Project access page which can be used by a project admin to specify roles to other users and
+	// restrict access within the project.
+	// +kubebuilder:validation:Optional
+	// +optional
+	ProjectAccess ProjectAccess `json:"projectAccess,omitempty"`
+}
+
+// ProjectAccess contains options for project access roles
+type ProjectAccess struct {
+	// availableClusterRoles are the list of ClusterRoles that are assignable to users
```
Comment (spadgett) [inline operator/v1/types_console.go:126]:
```suggestion
	// availableClusterRoles is the list of ClusterRole names that are assignable to users
```

File: operator/v1/types_console.go [hunk 4]
```diff
@@ -113,6 +113,21 @@ type ConsoleCustomization struct {
 	// +kubebuilder:validation:Optional
 	// +optional
 	DeveloperCatalog DeveloperConsoleCatalogCustomization `json:"developerCatalog,omitempty"`
+	// projectAccess allows customizing the available list of ClusterRoles in the Developer Console
+	// Project access page which can be used by a project admin to specify roles to other users and
+	// restrict access within the project.
```
Comment (spadgett) [inline operator/v1/types_console.go:118]:
```suggestion
	// projectAccess allows customizing the available list of ClusterRoles in the Developer perspective
	// Project access page which can be used by a project admin to specify roles to other users and
	// restrict access within the project. If set, the list will replace the default ClusterRole options.
```

**Pull request merged**